src/go/configmanager/config_manager.go

```diff
@@ -154,64 +146,52 @@ func NewConfigManager(mf *metadata.MetadataFetcher, opts options.ConfigGenerator
-		m.serviceName, m.curConfigID, rolloutStrategy)
+		m.serviceName, m.curConfigId(), rolloutStrategy)
 
 	if rolloutStrategy == util.ManagedRolloutStrategy {
-		go func() {
-			glog.Infof("start checking new rollouts every %v seconds", *checkNewRolloutInterval)
-			m.checkRolloutsTicker = time.NewTicker(*checkNewRolloutInterval)
-			for range m.checkRolloutsTicker.C {
-				m.Infof("check new rollouts for service %v", m.serviceName)
-				// only log error and keep checking when fetching rollouts and getting newest config fail
-				newRolloutID, newConfigID, err := loadConfigFromRollouts(m.serviceName, m.curRolloutID, m.curConfigID, mf)
-				if err != nil {
-					glog.Errorf("error occurred when checking new rollouts, %v", err)
-				}
-				if m.curRolloutID != newRolloutID && m.curConfigID != newConfigID {
-					m.curRolloutID = newRolloutID
-					m.curConfigID = newConfigID
-					if err := m.updateSnapshot(); err != nil {
-						glog.Errorf("error occurred when checking new rollouts, %v", err)
-					}
-				}
+		m.serviceConfigFetcher.SetFetchConfigTimer(checkNewRolloutInterval, func(serviceConfig *confpb.Service) {
+			err := m.applyServiceConfig(serviceConfig)
+			if err != nil {
+				glog.Errorf("error occurred when checking new rollouts, %v", err)
 			}
-		}()
+		})
 	}
 	return m, nil
 }
 
-// updateSnapshot should be called when starting up the server.
-// It calls ServiceManager Server to fetch the service configuration in order
-// to dynamically configure Envoy.
-func (m *ConfigManager) updateSnapshot() error {
-	serviceConfig, err := fetchConfig(m.serviceName, m.curConfigID, m.metadataFetcher)
+func (m *ConfigManager) readAndApplyServiceConfig(servicePath string) error {
+	config, err := ioutil.ReadFile(servicePath)
 	if err != nil {
-		return fmt.Errorf("fail to fetch service config, %s", err)
+		return fmt.Errorf("fail to read service config file: %s, error: %s", servicePath, err)
 	}
 
-	return m.applyServiceConfig(serviceConfig)
-}
-
-func (m *ConfigManager) readAndApplyServiceConfig(servicePath string) error {
-	serviceConfig, err := readConfig(servicePath)
+	serviceConfig, err := util.UnmarshalServiceConfig(bytes.NewReader(config))
 	if err != nil {
-		return fmt.Errorf("fail to read service config file: %s, error: %s", servicePath, err)
+		return fmt.Errorf("fail to unmarshal service config: %v, error: %s", config, err)
 	}
+
 	m.serviceName = serviceConfig.GetName()
-	m.curConfigID = serviceConfig.GetId()
+	m.configIdServiceConfigFile = serviceConfig.GetId()
 
 	return m.applyServiceConfig(serviceConfig)
 }
 
 func (m *ConfigManager) applyServiceConfig(serviceConfig *confpb.Service) error {
 	var err error
-	m.serviceInfo, err = configinfo.NewServiceInfoFromServiceConfig(serviceConfig, m.curConfigID, m.envoyConfigOptions)
+	m.serviceInfo, err = configinfo.NewServiceInfoFromServiceConfig(serviceConfig, m.curConfigId(), m.envoyConfigOptions)
```

Comment:
serviceConfig should have id() field which is the config_id.  not need to call m.curConfigId() to get it.